python/mspasspy/db/database.py

```diff
@@ -574,45 +645,182 @@ def save_data(self, mspass_object, mode="promiscuous", storage_mode='gridfs', di
+
+            # FIXME starttime will be automatically created in this function
+            self._sync_metadata_before_update(mspass_object)
 
-        update_res_code = -1
+            # This method of Metadata returns a list of all
+            # attributes that were changed after creation of the
+            # object to which they are attached.
+            changed_key_list=mspass_object.modified()
+
+            copied_metadata = Metadata(mspass_object)
+
+            # clear all the aliases
+            #TODO  check for potential bug in handling clear_aliases
+            # and modified method - i.e. keys returned by modified may be
+            # aliases
+            save_schema.clear_aliases(copied_metadata)
+
+            # remove any defined items in exclude list
+            for k in exclude_keys:
+                if k in copied_metadata:
+                    copied_metadata.erase(k)
+            # Now remove any readonly data
+            for k in copied_metadata.keys():
+                if save_schema.is_defined(k):
+                    if save_schema.readonly(k):
+                        if k in changed_key_list:
+                            newkey='READONLYERROR_'+k
+                            copied_metadata.change_key(k,newkey)
+                            mspass_object.elog.log_error('Database.save_data',
+                                'readonly attribute with key='+k+' was improperly modified.  Saved changed value with key='+newkey,
+                                ErrorSeverity.Complaint)
+                        else:
+                            copied_metadata.erase(k)
+            # Done editing, now we convert copied_metadata to a python dict
+            # using this Metadata method or the long version when in cautious or pedantic mode
+            insertion_dict = dict()
+            if mode == "promiscuous":
+                # A python dictionary can use Metadata as a constructor due to
+                # the way the bindings were defined
+                insertion_dict = dict(copied_metadata)
+            else:
+                # Other modes have to test every key and type of value
+                # before continuing.  pedantic kills data with any problems
+                # Cautious tries to fix the problem first
+                # Note many errors can be posted - one for each problem key-value pair
+                for k in copied_metadata:
+                    if save_schema.is_defined(k):
+                        if not isinstance(copied_metadata[k], save_schema.type(k)):
+                            insertion_dict[k]=copied_metadata[k]
+                        else:
+                            if mode == 'pedantic':
+                                mspass_object.kill()
+                                message ='pedantic mode error:  key='+k
+                                if not save_schema.is_defined(k):
+                                    message += ' not defined in schema.'
```

Comment:
This condition checking is redundant because, at line 703, it's always guaranteed that the key is defined.